Add deduplication logic

PaliC opened the pull request:
This pull request introduces deduplication functionality to the `export.py` script and updates the documentation to include testing instructions. The key changes include integrating a deduplication module, handling deduplicated datasets, and providing a detailed test setup for verifying the deduplication logic.

The deduplication functionality is specifically does 3 things
1) Filter out only for the columns `['code', 'run_mode', 'run_passed', 'run_meta', 'submission_id']` to make processing manageable.  Otherwise, I was running into issues with pandas loading the submissions dataframe.
2) It bins the submissions by success, run mode, and score (if available) or duration of the run. We then deduplicate within these bins as minhash takes a while (deduping this dataset took an hour)
3) We first dedup using a hash
4) We then dedup using minhash lsh (this is what takes a while).  You can find an explanation of this process here https://medium.com/@omkarsoak/from-min-hashing-to-locality-sensitive-hashing-the-complete-process-b88b298d71a1

There is also a good amount of tests added to make sure this stuff actually works

PaliC committed:
Add deduplication logic

PaliC committed:
Add deduplication logic

PaliC committed:
magic number removal

PaliC committed:
magic number removal

msaroufim commented:
Thanks! Mind sharing some more stats on the real dataset?
1. How much data was around before your change
2. How much data is around after
3. Some manual vibe checks as well of kernels that were filtered out would be nice to sanity check

Code is quite long to review but at least above should give us some more confidence before merge

PaliC committed:
Add deduplicated datasets

PaliC committed:
Add deduplicated datasets

ngc92 commented on dedup.py:
```diff
@@ -0,0 +1,480 @@
+                        # For leaderboard mode with successful runs, prefer higher scores
+                        if run_mode == 'leaderboard' and row.get('run_passed') == True:
+                            if row.get('run_score', 0) > existing_row.get('run_score', 0):
+                                unique_entries[content_hash] = row
```
I think scores are still lower = better; run.duration is the end-to-end wallclock time for the entire run, including, e.g., testing code, whereas score is the geomean of all benchmarks

PaliC replied:
oops, I see. I'll rerun things and reupload

PaliC commented:
@msaroufim for the first two questions we have these

✓ Loaded submissions.parquet: 40,095 entries
✓ Loaded deduplicated_submissions.parquet: 15,238 entries

============================================================
COUNT COMPARISON
============================================================
Original submissions:      40,095
Deduplicated submissions:  15,238
Removed entries:           24,857
Percentage removed:        62.00%

So it seems like a lot of the kernels are actually similar. I'll rerun things and grab some files for a sanity check

PaliC commented:
@msaroufim https://www.diffchecker.com/KamzTAeT/ (I think this one is more similar) and https://www.diffchecker.com/MtK5pbWL/ show a entry which was removed due to depulicaton (on right) compared to two other entries that remained in the dataset. We can be a bit less aggressive with deduping, as they look sortof different.

Commit:
Updates for 2nd comptetition

This change modifies the extraction processes so that it uses a lot less
memory. In particular, the process no longer loads the whole dataset
into memory before exporting to parquet files. Instead, it processes the
dataset into small, incremental parquet files, and then consolidates
these files into a single file as the final step.

b9r5 committed:
Updated extraction scripts for 2nd competition

This change modifies the extraction processes so that it uses a lot less memory. In particular, the process no longer loads the whole dataset into memory before exporting to parquet files. Instead, it processes the dataset into small, incremental parquet files, and then consolidates these files into a single file as the final step.

Commit:
remove test

Commit:
Merge branch 'dedup' of github.com:PaliC/kernelbot-data into dedup

# Conflicts:
#	export.py

PaliC commented:
Updated values
Successful submissions
```bash
Deduplication results Summary:
Original rows: 60357
After hash based dedup dedup: 22718 rows
Final rows: 9281
Removed 51076 duplicates (84.6%)
Saved to data/successful_submissions_deduplicated.parquet
```

Submissions
```bash
Flattening and saving...                                                                                                     
Deduplication results Summary:
Original rows: 109709
After hash based dedup dedup: 47362 rows
Final rows: 19012
Removed 90697 duplicates (82.7%)
Saved to data/submissions_deduplicated.parquet
```

Commit:
update
